### PR TITLE
ScriptEngine Threading Repairs

### PR DESCRIPTION
**Pull request by ZappoMan**
Fixes the threading issues with Entity Scripts and other users of ScriptEngine.

* Moves the Agent specific code out of ScriptEngine and into Agent.
* Major cleanup on the public interface to ScriptEngine
* wrap all public interfaces in thread-safe invokeMethod wrappers
* change all access to script resources to assume asynchronous loading

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 94b273a029033da34526f9a9e3d829fbb0f8e258 available at: https://gist.github.com/6226fb87a060eb8c90a9<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5804.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5804.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 54a45f317387ed5ef6fecedd827b9d2dffbca920 available at: https://gist.github.com/6226fb87a060eb8c90a9<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5804.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5804.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: e92d195030b2188ecc194e78e18ad80dc47637b7 available at: https://gist.github.com/6226fb87a060eb8c90a9<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5804.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5804.dmg

File: libraries/entities-renderer/src/EntityTreeRenderer.cpp [hunk 1]
```diff
@@ -1042,28 +770,16 @@ void EntityTreeRenderer::addEntityToScene(EntityItemPointer entity) {
...
-}
-
-void EntityTreeRenderer::checkAndCallUnload(const EntityItemID& entityID) {
-    if (_tree && !_shuttingDown) {
-        QScriptValue entityScript = getPreviouslyLoadedEntityScript(entityID);
-        if (entityScript.property("unload").isValid()) {
-            QScriptValueList entityArgs = createEntityArgs(entityID);
-            entityScript.property("unload").call(entityScript, entityArgs);
+        EntityItemPointer entity = getTree()->findEntityByEntityItemID(entityID);
+        if (!entity->getScript().isEmpty()) {
```
Comment (jherico) [inline libraries/entities-renderer/src/EntityTreeRenderer.cpp]: should be `if (entity && !entity->getScript().isEmpty())`

File: libraries/networking/src/ResourceManager.cpp [hunk 2]
```diff
@@ -17,21 +17,31 @@
 
 #include <SharedUtil.h>
 
-ResourceRequest* ResourceManager::createResourceRequest(QObject* parent, const QUrl& url) {
+QUrl ResourceManager::normalizeURL(const QUrl& url) {
     auto scheme = url.scheme();
+    if (!(scheme == URL_SCHEME_FILE ||
```
Comment (jherico) [inline libraries/networking/src/ResourceManager.cpp]: what, no gopher?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 70d2288407816cea1cf1862af7e211b6f00dd0cf available at: https://gist.github.com/6226fb87a060eb8c90a9<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5804.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5804.dmg

File: libraries/networking/src/ResourceManager.cpp (hunk 2 above)
Comment (ZappoMan) [inline libraries/networking/src/ResourceManager.cpp]: I almost added finger support.

**Pull request merged**